### PR TITLE
Modify the handleStepOption method

### PR DESCRIPTION
**Pull request by sahandilshan**
## Purpose
* Modify the `handleStepOptions` method to work with `markAsSubjectIdentifierStep` and `markAsSubjectAttributeStep` attributes.
* Fixed https://github.com/wso2/product-is/issues/12896

With this implementation, we can set the **Subject Identifier Step** and **Subject Attribute Step** with the adaptive authentication script as shown below,

Screenshot 2022-01-04 at 12 41 01 PM

### CODE REVIEW

**Comment by mefarazath** [comment]
@sahandilshan can you link the related issue?

**Comment by pulasthi7** [comment]
Can you please explain what you did and any example use case/script or something as well (either here or in the issue itself). Also, I guess this has doc impact as well?

**Comment by sahandilshan** [comment]
@mefarazath & @pulasthi7, I have created an issue in the product-is repo and linked that issue. As @pulasthi7 requested that issue was explained with a proper use case.

**Comment by sahandilshan** [comment]
Relevant docs update for this modification can be found here https://github.com/wso2/docs-is/pull/2828

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkConstants.java [hunk 1]
```diff
@@ -500,6 +500,8 @@ private RequestAttribute() {
         public static final String STEP_OPTIONS = "stepOptions";
         public static final String AUTHENTICATOR_PARAMS = "authenticatorParams";
         public static final String FORCE_AUTH_PARAM = "forceAuth";
+        public static final String SUBJECT_IDENTIFIER = "subjectIdentifier";
+        public static final String SUBJECT_ATTRIBUTE_STEP = "subjectAttributeStep";
```
Comment (senthalan) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkConstants.java:504]: Do we need to have the step word here?

 As this is under stepOptions won't the "subjectAttribute" enough?

Comment (sahandilshan) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkConstants.java]: This has been fixed with fa17986bf0651c6fb99754285ab171f2d38f6173

**Review by pulasthi7 — APPROVED** [review]

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java [hunk 2]
```diff
@@ -1012,6 +1018,24 @@ private JSExecutionMonitorData endScriptExecutionMonitor(String identifier) {
         return getJSExecutionSupervisor().completed(identifier);
     }
 
+    private void setCurrentStepAsSubjectIdentifier(StepConfig stepConfig) {
+        stepNamedMap.forEach((integer, config) -> { // remove existing subject identifier step
```
Comment (senthalan) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java:1022]: A new line before the method body.

Comment (senthalan) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java:1022]: Comments need to start with capital letter and ends with full stop

Comment (sahandilshan) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java:1022]: Fixed with https://github.com/wso2/carbon-identity-framework/pull/4068

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java [hunk 3]
```diff
@@ -1012,6 +1018,24 @@ private JSExecutionMonitorData endScriptExecutionMonitor(String identifier) {
...
+            if (config.isSubjectIdentifierStep()) {
+                config.setSubjectIdentifierStep(false);
+            }
+        });
+        stepConfig.setSubjectIdentifierStep(true);
+    }
+
+    private void setCurrentStepAsSubjectAttribute(StepConfig stepConfig) {
```
Comment (piraveena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java:1030]: Add new line after method signature

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java [hunk 4]
```diff
@@ -1012,6 +1018,24 @@ private JSExecutionMonitorData endScriptExecutionMonitor(String identifier) {
...
+        stepNamedMap.forEach((integer, config) -> { // remove existing subject attribute step
```
Comment (piraveena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java:1031]: `Remove existing subject attribute step.`

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java (hunk 3 above)
Comment (sahandilshan) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java:1030]: This has already been fixed with https://github.com/wso2/carbon-identity-framework/pull/4068/files

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java (hunk 4 above)
Comment (sahandilshan) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java:1031]: This has already been fixed with https://github.com/wso2/carbon-identity-framework/pull/4068/files

**Pull request merged**